File: gatling-jms/src/main/scala/io/gatling/jms/jndi/JmsJndiConnectionFactoryBuilder.scala
/*
 * Copyright 2011-2025 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.jms.jndi

import java.util.{ Hashtable => JHashtable }
import javax.naming.{ Context, InitialContext }

import scala.jdk.CollectionConverters._

import io.gatling.commons.model.Credentials

import com.typesafe.scalalogging.StrictLogging
import jakarta.jms.ConnectionFactory

object JmsJndiConnectionFactoryBuilder {
  object Base {
    def connectionFactoryName(cfn: String): Url = new Url(cfn)
  }

  final class Url(connectionFactoryName: String) {
    def url(theUrl: String): ContextFactory =
      ContextFactory(connectionFactoryName, theUrl, None, Map.empty)
  }

  final case class ContextFactory(
      connectionFactoryName: String,
      url: String,
      credentials: Option[Credentials],
      properties: Map[String, String]
  ) {
    def credentials(user: String, password: String): ContextFactory =
      copy(credentials = Some(Credentials(user, password)))

    def property(key: String, value: String): ContextFactory =
      copy(properties = properties.updated(key, value))

    def contextFactory(cf: String): JmsJndiConnectionFactoryBuilder =
      new JmsJndiConnectionFactoryBuilder(cf, connectionFactoryName, url, credentials, properties)
  }
}

final class JmsJndiConnectionFactoryBuilder(
    contextFactory: String,
    connectionFactoryName: String,
    url: String,
    credentials: Option[Credentials],
    jndiProperties: Map[String, String]
) extends StrictLogging {
  def build(): ConnectionFactory = {
    // create InitialContext
    val properties = new JHashtable[String, String]
    properties.put(Context.INITIAL_CONTEXT_FACTORY, contextFactory)
    properties.put(Context.PROVIDER_URL, url)

    credentials.foreach { credentials =>
      properties.put(Context.SECURITY_PRINCIPAL, credentials.username)
      properties.put(Context.SECURITY_CREDENTIALS, credentials.password)
    }
    properties.putAll(jndiProperties.asJava)

    val ctx = new InitialContext(properties)
    logger.debug(s"Got InitialContext $ctx")

    // create QueueConnectionFactory
    val qcf = ctx.lookup(connectionFactoryName).asInstanceOf[ConnectionFactory]
    logger.debug(s"Got ConnectionFactory $qcf")
    qcf
  }
}

File: gatling-jms-java/src/main/java/io/gatling/javaapi/jms/JmsJndiConnectionFactoryBuilder.java
/*
 * Copyright 2011-2022 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.javaapi.jms;

import javax.annotation.Nonnull;
import javax.jms.ConnectionFactory;

/**
 * DSL for building JNDI ConnectionFactory configurations
 *
 * <p>Immutable, so all methods return a new occurrence and leave the original unmodified.
 */
public final class JmsJndiConnectionFactoryBuilder {

  private final io.gatling.jms.jndi.JmsJndiConnectionFactoryBuilder wrapped;

  private JmsJndiConnectionFactoryBuilder(
      io.gatling.jms.jndi.JmsJndiConnectionFactoryBuilder wrapped) {
    this.wrapped = wrapped;
  }

  public ConnectionFactory build() {
    return wrapped.build();
  }

  public static final class Base {

    public static final Base INSTANCE = new Base();

    private Base() {}

    /**
     * Define the ConnectionFactory JNDI name
     *
     * @param cfn the name
     * @return the next DSL step
     */
    public Url connectionFactoryName(String cfn) {
      return new Url(io.gatling.jms.Predef.jmsJndiConnectionFactory().connectionFactoryName(cfn));
    }
  }

  public static final class Url {
    private final io.gatling.jms.jndi.JmsJndiConnectionFactoryBuilder.Url wrapped;

    private Url(io.gatling.jms.jndi.JmsJndiConnectionFactoryBuilder.Url wrapped) {
      this.wrapped = wrapped;
    }

    /**
     * Define the ConnectionFactory url
     *
     * @param url the url
     * @return the next DSL step
     */
    @Nonnull
    public ContextFactory url(@Nonnull String url) {
      return new ContextFactory(wrapped.url(url));
    }
  }

  public static final class ContextFactory {
    private final io.gatling.jms.jndi.JmsJndiConnectionFactoryBuilder.ContextFactory wrapped;

    private ContextFactory(
        io.gatling.jms.jndi.JmsJndiConnectionFactoryBuilder.ContextFactory wrapped) {
      this.wrapped = wrapped;
    }

    /**
     * Define the credentials to populate the {@link javax.naming.Context#SECURITY_PRINCIPAL} and
     * {@link javax.naming.Context#SECURITY_CREDENTIALS} properties.
     *
     * @param user the user
     * @param password the password
     * @return a new instance
     */
    @Nonnull
    public ContextFactory credentials(@Nonnull String user, @Nonnull String password) {
      return new ContextFactory(wrapped.credentials(user, password));
    }

    /**
     * Define some property
     *
     * @param key the property key
     * @param value the property value
     * @return a new instance
     */
    @Nonnull
    public ContextFactory property(@Nonnull String key, @Nonnull String value) {
      return new ContextFactory(wrapped.property(key, value));
    }

    /**
     * Define the ConnectionFactory class
     *
     * @param clazz the class
     * @return a usable {@link JmsJndiConnectionFactoryBuilder}
     */
    @Nonnull
    public JmsJndiConnectionFactoryBuilder contextFactory(@Nonnull String clazz) {
      return new JmsJndiConnectionFactoryBuilder(wrapped.contextFactory(clazz));
    }
  }
}
